### ISSUE TITLE
Intermittent container start failures "iptables: No chain/target/match by that name"

### ISSUE THREAD

**Issue by leth** [docker/docker#14464]
#### Description of problem:

We regularly see the following error from docker in our test environment when we try to start a bunch of linked containers in quick succession:
```
Cannot start container ffd703e4880036a7ee2bbe977ba1da96053cb78d7b5cbeec701548c345a1b49b:
iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 5432 -j DNAT --to-destination 10.0.0.16:5432 ! -i docker0:
iptables: No chain/target/match by that name.
```
(wrapped for readability)

We are using docker-in-docker so that our project tools don't have to be installed on our jenkins machine. We see the error from the inner docker daemon.

#### `docker version`:
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64

#### `docker info`:
Containers: 6
Images: 169
Storage Driver: aufs
 Root Dir: /home/jenkins/workspace/ttvm-integration-tests/dev/build/docker/graph/aufs
 Backing Filesystem: extfs
 Dirs: 181
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-55-generic
Operating System: Ubuntu 14.04.2 LTS (containerized)
CPUs: 8
Total Memory: 7.798 GiB
Name: 8245cdb1ddbf
ID: AEAG:ZDB4:JESL:N7Y6:VPRG:7K3D:O5TY:DQZY:VRZT:FIDE:ATI3:3M24
WARNING: No swap limit support

#### `uname -a`:
Linux 8245cdb1ddbf 3.13.0-55-generic #94-Ubuntu SMP Thu Jun 18 00:27:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):
Vmware VM, jenkins build node, running docker inside docker
NB: we are starting containers on the inner docker instance when we see the issue

How reproducible:
50% of the time, if not more

#### Steps to Reproduce:
1. Start linked containers
2.
3.

#### Actual Results:
500 Internal Server Error, with message:
```
Cannot start container ffd703e4880036a7ee2bbe977ba1da96053cb78d7b5cbeec701548c345a1b49b: iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 5432 -j DNAT --to-destination 10.0.0.16:5432 ! -i docker0: iptables: No chain/target/match by that name.
```

#### Expected Results:
No errors

**Comment by cpuguy83** [comment]
ping @mrjana

**Comment by leth** [comment]
Okay, I can rule out the docker-in-docker part of this. I'm running things locally and still see the same failure:
```
500 Server Error: Internal Server Error ("Cannot start container d8d4a837885991c5eccbb25dc2d6d234635e923931b3d749ef0e6432a3198a5d: iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 5002 -j DNAT --to-destination 172.17.0.75:5002 ! -i docker0: iptables: No chain/target/match by that name.
```

```
$ uname -a
Linux kohlenstoff 3.13.0-24-generic #47-Ubuntu SMP Fri May 2 23:30:00 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
                                                                                                                                                               $ docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64

$ docker info
Containers: 31
Images: 206
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 268
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-24-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 4
Total Memory: 7.681 GiB
Name: kohlenstoff
ID: A2Z5:S5HC:B3ZL:AI4R:TSKW:3HTP:5K4D:22GZ:BTZL:AV4T:ZA73:BRKO
WARNING: No swap limit 
```

**Comment by mrjana** [comment]
@leth Is it possible that you are running multiple docker daemon instances either in d-in-d or without? Because this is simply not possible even theoretically otherwise because the only time we remove the DOCKER chain is at the startup of the daemon(just once) and there is no way we remove it after.  The only other possibility is if you are running any other firewall daemons which are cleaning up iptable rules. Also can you try 1.7.1 to see if the issue is still there?

**Comment by leth** [comment]
Our environment is a bit complex; I'll try and describe it:
```
Jenkins slave host
- docker
  - test_job_a container
     - docker
        - service_a container --net=host --bridge=my_bridge 
        - service_b container
        - ...

  - test_job_b container
     - docker
        - service_a container  --net=host --bridge=my_bridge
        - service_b container
        - ...
```

The jenkins slave host is probably still on 1.6.x, the test job containers will be (now) running the latest release.
I just looked through the last 30 builds, and I've not seen this issue happening since 13-Jul-2015 11:46am which was build now - 27.

**Comment by mrjana** [comment]
@leth Interesting. Did you update inside docker to bleeding edge or something? Any theories on why stopped seeing the problem?

**Comment by leth** [comment]
Nope, no theories yet! No, the inner dockers are just built from the normal apt repo. I'll let you know if I see it again, and if not close this soon.

**Comment by leth** [comment]
I just saw it again (I removed my retry-based workaround); I have a hunch that it's when `test_job_a` and `test_job_b` are starting up the inner containers concurrently. does that sound possible?

**Comment by leth** [comment]
I've updated the jenkins slave host, so the outer docker is now 1.7.1

**Comment by mrjana** [comment]
@leth Do you have firewalld enabled by any chance in your host?

**Comment by leth** [comment]
Oh, sorry, I messed up the diagram a bit:

```
Jenkins slave host
- docker
  - test_job_a container
     - docker
        - service_a container --net=host 
           - docker --bridge=my_bridge 
        - service_b container
        - ...

  - test_job_b container
     - docker
        - service_a container  --net=host
           - docker --bridge=my_bridge
        - service_b container
        - ...
```

I don't think so, can't see anything for it in `/etc/init/` or `/etc/init.d/` - the build slave is ubuntu:14.04.

**Comment by leth** [comment]
It's only the service_a instances which start docker inside them.
We see the problem when starting the service_* containers, some of which are linked together.

**Comment by mrjana** [comment]
@leth ok this makes sense a bit. You are starting service_ containers as --net=host and starting docker daemons inside them. So  one docker daemon may be deleting the DOCKER chain while it's starting up while your other docker daemon might be starting a container and linking and so trying to use the DOCKER chain at the same time

**Comment by leth** [comment]
Right. Is there any way to avoid the name clash? Presumably the problem is that there's no namespacing on iptables yet? 

On 17 July 2015 14:13:45 GMT-05:00, Jana Radhakrishnan <notifications@github.com> wrote:
>@leth ok this makes sense a bit. You are starting service_ containers
>as --net=host and starting docker daemons inside them. So  one docker
>daemon may be deleting the DOCKER chain while it's starting up while
>your other docker daemon might be starting a container and linking and
>so trying to use the DOCKER chain at the same time
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/docker/docker/issues/14464#issuecomment-122377845

-- 
Sent from my Phone. Please excuse my brevity.

**Comment by leth** [comment]
I should clarify our intent here; we're trying to isolate everything inside test_a from test_b and vice-versa.
It's only when we're when deployed, that we want service_a's `--net host` to truly be the host's networking adapter.
We're using a custom bridge so that service_a's docker container IP allocation does not clash with our service IPs as a result of the `--net host`

**Comment by leth** [comment]
I've made `test_job_a` and `test_job_b` mutually exclusive, we now see the issue less, but we still see it.

**Comment by skozin** [comment]
I am seeing the same error in Docker 1.8.1 very frequently:

```
Cannot start container 6eaa4bc1d3e9a32291a52c19afe68ab054c68b551eb1677c5534b591f37980f4:
iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 32802 -j DNAT
                 --to-destination 10.0.42.47:9999 ! -i docker0:
iptables: No chain/target/match by that name.
```

I am running my containers _without_ `--net host` and using the default bridge `docker0` (with `--bip 10.0.42.1/24`). Attempting to start the same container again usually helps.

```
docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:40:42 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:40:42 UTC 2015
 OS/Arch:      linux/amd64
```

**Comment by dgromov** [comment]
Is there any work around for this? I'm seeing this with both 1.8.1 and 1.9.0-Dev. I'm also seeing this frequently. I'm seeing these issues running as part of a mesos executor.

```
> docker version

Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   7f8954c
 Built:        Tue Sep  1 22:19:33 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   7f8954c
 Built:        Tue Sep  1 22:19:33 UTC 2015
 OS/Arch:      linux/amd64
```

```
> docker info

Containers: 19
Images: 124
Storage Driver: aufs
 Root Dir: /data/docker/aufs
 Backing Filesystem: extfs
 Dirs: 162
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-46-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 8
Total Memory: 29.45 GiB
Name: xxx
ID: BAZH:BBPT:GLWS:2CVJ:RK2K:E3MR:RGIZ:SJTB:LZEH:BPCN:YUOR:CEFV
WARNING: No swap limit support
```

I was on docker 1.6.0 previously using the same kernel.

**Comment by mrjana** [comment]
@dgromov This should get fixed in 1.8.2. Please test 1.8.2 RC1(whenever it's out) and provide feedback.

**Comment by dgromov** [comment]
@mrjana thank you, will do. Just to confirm, has the fix not been merged to the master branch? The 1.9.0-Dev binary I'm using comes from https://master.dockerproject.org/linux/amd64/docker.

**Comment by mrjana** [comment]
@dgromov Since the libnetwork code has diverged a lot b/w 1.8 release and master we are doing this in two steps. 1.8.2 is currently being merged. We will merge that in master as well shortly

**Comment by crosbymichael** [comment]
This has been fixed by #16038

thanks!

**Comment by leth** [comment]
Do you have a test for this issue?
